### PR TITLE
Convert all ObjectIDs Convert all ObjectIDs

### PR DESCRIPTION
**Pull request by kimkha**
Please review.

#### Related issues

- #363 RoleMapping not work well with mongoDB connector

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

File: lib/mongodb.js [hunk 1]
```diff
@@ -363,6 +363,26 @@ MongoDB.prototype.coerceId = function(model, id) {
 };
 
 /**
+ * Convert all objectIds
```
Comment (b-admike) [inline lib/mongodb.js]: Here you can say something like `coerce type of all object id properties to mongoDB's ObjectID`

**Comment by rmg** [comment]
@slnode ok to test

**Comment by b-admike** [comment]
@kimkha thank you for your contribution. code LGTM; please add some unit tests for the new function you wrote and to avoid regressions.

**Comment by b-admike** [comment]
connected to #363

File: lib/mongodb.js [hunk 2]
```diff
@@ -363,6 +363,26 @@ MongoDB.prototype.coerceId = function(model, id) {
...
+ *
+ * @param {String} model The model name
+ * @param {Object} data The model data
+ * @returns {Object} The new model data
+ */
+MongoDB.prototype.convertAllObjectID = function(model, data) {
```
Comment (b-admike) [inline lib/mongodb.js]: Please ignore. It is the other way around and you have fixed it :-)

**Comment by b-admike** [comment]
@kimkha thanks for applying the feedback. If you need help with writing a test case for this PR, let me know.

**Comment by kimkha** [comment]
@b-admike I'm new with mocha, could you please write a test for me? My source code just convert String to ObjectId before saving. So, I think it'll OK if we save string and DB present as ObjectId.

**Comment by bajtos** [comment]
@fabien Could you PTAL? I think this may go against the gist of your `strictObjectIDCoercion` setting.

I am also concerned that the change proposed by this pull request will break backwards compatibility. Data stored with the current version of the connector will store ObjectID-like properties differently than the data stored with the new version of the connector. As a result, people may end up with inconsistent data format in their databases.

Thoughts?

**Comment by fabien** [comment]
In principle, this should be fine as long as `isObjectIDProperty` is used (which follows the `strictObjectIDCoercion` as well). As such, backwards compatibility would be maintained. But it's hard to know what's out there in the wild, of course.

**Comment by fabien** [comment]
That said, wouldn't it make sense to add this logic to `toDatabase` instead of introducing a new method that is nearly always called alongside it?

**Comment by b-admike** [comment]
@fabien I tested the fix in this PR against the original issue #363 and my conclusion after testing with the flag `RoleMapping.settings.strictObjectIDCoercion` in the code is that it doesn't matter what you set the flag to if you are using this patch. Thus, I agree with what @bajtos said about this PR going against the gist of the `strictObjectIDCoersion` flag and backwards compatibility. The issue #363 is fixed if we use `RoleMapping.settings.strictObjectIDCoercion = true` in the boot script for the `loopback-example-access-control` application. I would like to propose that we document what `strictObjectIDCoersion` flag does so that users are aware of it and know when to use it. @kimkha yes I can write the test, but I think we should figure out how to proceed further with this PR first.

**Comment by kimkha** [comment]
OK, actually, I don't understand what all you mention... 😄 I'm new with loopback, just find an issue and I think I can help to fix it... So please discuss each other and let me know when you finish...

**Comment by b-admike** [comment]
@kimkha what I am saying is that #363 works when we use `RoleMapping.settings.strictObjectIDCoercion = true` without your fix: see https://github.com/strongloop/loopback-connector-mongodb/issues/363#issuecomment-297437024. Can you give it a try?

**Comment by kimkha** [comment]
@b-admike Oh, `strictObjectIDCoercion = true` works nice. So that mean we don't need to fix with this PR, right? But we should make doc clearer, I think.

**Comment by b-admike** [comment]
@kimkha yeah you're right and that is my proposal as well. I have filed https://github.com/strongloop/loopback-connector-mongodb/issues/371. You can close this PR if you are satisfied and we can always revisit if necessary. Thank you for your contribution!

**Pull request closed without merging**